### PR TITLE
Helm Release Pipeline

### PR DESCRIPTION
**Pull request by naemono**
This adds a Buildkite pipeline for releasing of ECK Helm charts.
This also adds a custom Helm release tool.

Included:

1. Pipeline for releasing ECK resource Helm charts.
2. Pipeline for releasing ECK Operator Helm Chart.
3. Go tool to handle releasing Helm Charts to Elastic Helm Buckets/Repositories.

What's been manually tested:

* Releasing of ECK resources Helm charts to both dev and production Elastic Helm Buckets/Repositories. (resolving eck-stack chart dependency issue)
* Releasing of ECK Operator Helm chart to dev Elastic Helm Bucket/Repository.

In progress:

- [x] Currently testing Buildkite integration (will fully test during actual release).

### CODE REVIEW

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:21]:
```suggestion
	bucketFlag          = "bucket"
```

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:61]:
```suggestion
			log.Printf("Releasing charts in (%s) to bucket (%s) in repo (%s)\n", viper.GetString(chartsDirFlag), bucket, chartsRepoURL)
```

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:64]:
```suggestion
					ChartsDir:           viper.GetString(chartsDirFlag),
```

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:67]:
```suggestion
					CredentialsFilePath: viper.GetString(credentialsFileFlag),
```

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:68]:
```suggestion
					DryRun:              viper.GetBool(dryRunFlag),
```

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:69]:
```suggestion
					Excludes:            viper.GetStringSlice(excludesFlag),
```

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:93]:
```suggestion
		"/tmp/credentials.json",
```

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:114]: I would enable vault by default.
```suggestion
		true,
```

File: hack/helm/release/cmd/root.go
Comment (naemono) [reply hack/helm/release/cmd/root.go:21]: What's the intention here?  I see no changes suggested.

Comment (thbkrkr) [reply hack/helm/release/cmd/root.go:21]: Oops, the constant can be deleted, it is no longer used.

File: hack/helm/release/readme.md
Comment (thbkrkr) [inline hack/helm/release/readme.md:46]:
```suggestion
| `--credentials-file`| Full path to credentials file to use for GCS bucket.                                                           | `HELM_CREDENTIALS_FILE` | `"/tmp/credentials.json"`                               |
```

File: hack/helm/release/internal/helm/helm.go
Comment (thbkrkr) [inline hack/helm/release/internal/helm/helm.go:239]: Why do we shortcut in the middle of `uploadChartsAndUpdateIndex` to not process charts with dependencies? IIUC the index of the charts with deps will be invalid because we didn't update the index of the parent charts. But I wonder if it is really that bad? With dry-run we want to test as much as possible of the process and we upload anything.

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:109]: The trick with the `--char-dir` and the `--excludes` to only release the eck-resources Helm charts is a bit twisted I think. Maybe it would be cleaner if it's possible to move the eck-resources charts to a parent directory at the same level than the operator?

Instead of:
```
release --char-dir deploy/eck-operator
release --char-dir deploy --excludes eck-operator
```

We could do:
```
release --char-dir deploy/eck-operator
release --char-dir deploy/eck-resources
```

File: .buildkite/pipeline.yml
Comment (naemono) [reply .buildkite/pipeline.yml:118]: Agreed @thbkrkr. Only thought here is that we're going to be releasing helm charts on these tags for versions that the `release` process isn't complete, as, I believe, that process still lives in jenkins, right?  I think I'd prefer to comment this section out for now, and only enable this when we can wait for the release process in BK to be complete.  Thoughts?

File: hack/helm/release/internal/helm/helm.go
Comment (naemono) [reply hack/helm/release/internal/helm/helm.go]: Ok, I guess that makes sense.  It's been removed.

File: .buildkite/pipeline.yml
Comment (naemono) [inline .buildkite/pipeline.yml:127]: This bothers me.  This seemed to be the `proper` way to do this via BK, but it's always failing in BK (which is 'expected'), which causes github 'checks' to always show failed.  I'm investigating another way to do this, or simply removing this `cd` step for now, and requiring manual release.  Thoughts?

Comment (naemono) [reply .buildkite/pipeline.yml]: removed per conversation with @thbkrkr

File: hack/helm/release/internal/helm/helm.go
Comment (thbkrkr) [inline hack/helm/release/internal/helm/helm.go:308]: We add Helm repositories even if there are no charts to download? Must be harmless I guess?

File: hack/helm/release/internal/helm/helm.go
Comment (thbkrkr) [inline hack/helm/release/internal/helm/helm.go:313]: The same message is already printed by `addHelmRepository(name, url)`.
```suggestion
```

File: hack/helm/release/internal/helm/helm.go
Comment (thbkrkr) [inline hack/helm/release/internal/helm/helm.go:346]: `name` and `url` arguments are not used. We always add "stable", "https://charts.helm.sh/stable" instead, normal?

File: hack/helm/release/internal/helm/helm.go
Comment (thbkrkr) [inline hack/helm/release/internal/helm/helm.go:289]: I've tested to release in `--env=dev` and I see that we are updating the 2 repositories, is it normal?:

```
2023/01/27 20:32:08 Updated repository index for (elastic) elastic
2023/01/27 20:32:08 Updated repository index for (elastic-dev) elastic-dev
```

File: .buildkite/scripts/helm/has-any-chartyml-changed.sh
Comment (thbkrkr) [inline .buildkite/scripts/helm/has-any-chartyml-changed.sh:1]: Let's remove it for now because it's no longer in use?

File: hack/helm/release/internal/helm/helm.go
Comment (thbkrkr) [inline hack/helm/release/internal/helm/helm.go:438]: Can we check this and remove the TODO?

File: .buildkite/pipeline-helm-release-eck-resources.yml
Comment (thbkrkr) [inline .buildkite/pipeline-helm-release-eck-resources.yml:19]: I think we can simplify a bit.
We can avoid the `cd` using `make -C`.
`VAULT_ROOT_PATH` isn't used, instead there is `HELM_VAULT_SECRET`. 
`HELM_CREDENTIALS_FILE` is set to `/tmp/credentials.jso` by default.
` --enable-vault` is true by default.
`--charts-dir` is set to `deploy` by default.

```suggestion
          - make -C hack/helm/release build 2> /dev/null
          - if [[ -n "$$HELM_BRANCH" ]]; then echo "checking out $$HELM_BRANCH branch" && git checkout $$HELM_BRANCH ; fi
          - hack/helm/release/bin/releaser --env=dev --excludes=eck-operator
```

File: hack/helm/release/Dockerfile
Comment (naemono) [reply hack/helm/release/Dockerfile]: Nope, removed.

File: hack/helm/release/internal/helm/helm.go
Comment (naemono) [reply hack/helm/release/internal/helm/helm.go:312]: I've updated this logic to break early.  It's harmless though.

File: hack/helm/release/internal/helm/helm.go
Comment (naemono) [reply hack/helm/release/internal/helm/helm.go:349]: This was fixed.  I was initially only adding stable, and then realized that for dependencies sake, itself would be needed.  Resolved.

File: hack/helm/release/internal/helm/helm.go
Comment (naemono) [reply hack/helm/release/internal/helm/helm.go]: Thanks for pointing this out.  I've verified, and since we're using constants now for dev/prod, we can assume we can use the `Path` parsed from "net/url".Parse, with the leading `/` trimmed.  👍

File: .buildkite/pipeline-helm-release-eck-resources.yml
Comment (naemono) [reply .buildkite/pipeline-helm-release-eck-resources.yml]: Excellent point.  Adjusted all pipelines.

File: hack/helm/release/readme.md
Comment (thbkrkr) [inline hack/helm/release/readme.md:1]: I'm a bit concerned about maintaining this readme with this high level of details. The tool already has a `--help` flag that covers a lot of information. We could stay more general.

File: hack/helm/release/cmd/root.go
Comment (thbkrkr) [inline hack/helm/release/cmd/root.go:57]: I think `--upload-index` has been removed.

```suggestion
		Example: fmt.Sprintf("  %s", "release --charts-dir=deploy --dry-run=false"),
```

File: .buildkite/pipeline.yml
Comment (thbkrkr) [inline .buildkite/pipeline.yml:133]: I think we said we will publish daily snapshots of all helm charts daily from `main` to the `dev` repository. There will be more change on this part when we introduce the pipeline-release.yml definition so I'm ok with merging the PR in this state and moving forward.

File: hack/helm/release/readme.md
Comment (thbkrkr) [inline hack/helm/release/readme.md:63]: What the interest of using this `HELM_BRANCH` environment variable instead of the regular required properties `commit` and `branch`?

File: hack/helm/release/internal/helm/helm.go
Comment (naemono) [reply hack/helm/release/internal/helm/helm.go:290]: I'm assuming you ran this locally?  It's loading your local configuration, which includes both elastic and elastic-dev repositories, and updating all the indexes, which is a harmless operation.  You won't see this in Buildkite.

```
func updateHelmRepositories() error {
	// simulates helm repo update
	f, err := repo.LoadFile(cli.New().RepositoryConfig)
	if err != nil {
		return fmt.Errorf("while loading repository file: %w", err)
	}
	for _, r := range f.Repositories {
		cr, err := repo.NewChartRepository(r, getter.All(cli.New()))
		if err != nil {
			return fmt.Errorf("while getting new chart repository: %w", err)
		}
		if _, err = cr.DownloadIndexFile(); err != nil {
			return fmt.Errorf("while updating repository index for (%s): %w", cr.Config.Name, err)
		}
		log.Printf("Updated repository index for (%s)\n", cr.Config.Name)
	}
	return nil
}
```

For me, it's even a bit more:
```
2023/01/30 13:58:20 Updated repository index for (grafana)
2023/01/30 13:58:21 Updated repository index for (metrics-server)
2023/01/30 13:58:22 Updated repository index for (prometheus-community)
2023/01/30 13:58:23 Updated repository index for (elastic)
2023/01/30 13:58:24 Updated repository index for (jetstack)
2023/01/30 13:58:24 Updated repository index for (cdf)
2023/01/30 13:58:25 Updated repository index for (argo)
2023/01/30 13:58:25 Updated repository index for (nginx-stable)
2023/01/30 13:58:25 Updated repository index for (aws-ebs-csi-driver)
2023/01/30 13:58:26 Updated repository index for (elastic-helm-dev)
```

File: hack/helm/release/readme.md
Comment (naemono) [reply hack/helm/release/readme.md:57]: @thbkrkr readme updated to try and explain this.  Let me know if it's more clear.

Comment (thbkrkr) [reply hack/helm/release/readme.md:57]: `branch` and `commit` properties select the branch from which to build the tool and `HELM_BRANCH` the branch from which to release? It's not clear for me why we couldn't just use `branch` and `commit` for the two. IIUC one interest to separate the two is to be able to release code where the tool didn't exist?

Comment (naemono) [reply hack/helm/release/readme.md:57]: I see your point. I guess the only possible use case for this after merge of this PR would be if changes were needed to the tooling. But if that was the case then we could just merge the tooling changes prior to running the tooling. I will simply remove this to make things more clear.

**Review by thbkrkr — APPROVED** [review]
LGTM

We will make small adjustments when integrating into the release pipeline.

**Pull request merged**
